<a href="https://colab.research.google.com/github/hoangvn111/Project-Analyzing-CIA-Factbook-Data-Using-SQL/blob/master/Project_Analyzing_CIA_Factbook_Data_Using_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [2]:
import sqlite3

conn = sqlite3.connect('/content/drive/MyDrive/my_datasets/Project_ Analyzing CIA Factbook Data Using SQL/factbook.db')
print("Opened database successfully")

Opened database successfully


# Introduction

In this project, we'll work with data from the [CIA World Factbook](https://www.cia.gov/the-world-factbook/), a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like the following:

* **population** — the global population.
* **population_growth** — the annual * * ?population growth rate, as a percentage.
* **area** — the total land and water area.

In [3]:
import pandas as pd
# Explore the database
df = pd.read_sql_query('''SELECT * 
                          FROM   facts
                          LIMIT  5;''', conn)
df

id code         name  ...  birth_rate  death_rate  migration_rate
0   1   af  Afghanistan  ...       38.57       13.89            1.51
1   2   al      Albania  ...       12.92        6.58            3.30
2   3   ag      Algeria  ...       23.67        4.31            0.92
3   4   an      Andorra  ...        8.13        6.96            0.00
4   5   ao       Angola  ...       38.78       11.49            0.46

[5 rows x 11 columns]

# Summary Statistics

Executing our previous query, we see the first few rows of our facts table look like this:

Here are the descriptions for some of the columns:

* **name** — the name of the country.
* **area—** the country's total area (both land and water).
* **area_land** — the country's land area in square kilometers.
* **area_water** — the country's waterarea in square kilometers.
* **population** — the country's population.
* **population_growth** — the country's population growth as a percentage.
* **birth_rate** — the country's birth rate, or the number of births per year per 1,000 people.
* death_rate — the country's death rate, or the number of death per year per 1,000 people.

In [4]:
# Write a single query that returns maximum polution
max_population = pd.read_sql_query('''SELECT MAX(population) 
                                        FROM facts;''', conn)
max_population

MAX(population)
0       7256490011

In [5]:
# Write a single query that returns minimum polution
min_population = pd.read_sql_query('''SELECT MIN(population) 
                                        FROM facts;''', conn)
min_population

MIN(population)
0                0

In [6]:
# Write a single query that returns maximum polution growth
max_population_growth = pd.read_sql_query('''SELECT MAX(population_growth) 
                                               FROM facts;''', conn)
max_population_growth

MAX(population_growth)
0                    4.02

In [7]:
# # Write a single query that returns minimum polution growth
min_population_growth = pd.read_sql_query('''SELECT MIN(population_growth) 
                                               FROM facts;''', conn)
min_population_growth

MIN(population_growth)
0                     0.0

# Exploring Outliers

We see a few interesting things in the summary statistics on the previous screen:

* There's a country with a population of 0
* There's a country with a population of 7256490011 (or more than 7.2 billion people)

In [8]:
# Write a query that returns the countries with the maximum population.
max_population = pd.read_sql_query('''SELECT *, MAX(population) 
                                        FROM facts;''', conn)
max_population

id code   name  ... death_rate migration_rate MAX(population)
0  261   xx  World  ...        7.8           None      7256490011

[1 rows x 12 columns]

In [9]:
# Write a query that returns the countries with the minimum population.
min_population = pd.read_sql_query('''SELECT *, MIN(population) 
                                        FROM facts;''', conn)
min_population

id code        name  ... death_rate  migration_rate MIN(population)
0  250   ay  Antarctica  ...       None            None               0

[1 rows x 12 columns]

In [10]:
# Write a single query that returns maximum polution excluding the row for the whole world
max_population = pd.read_sql_query('''SELECT *, MAX(population) 
                                        FROM facts
                                       WHERE name <> 'World';''', conn)
max_population

id code   name  ...  death_rate  migration_rate  MAX(population)
0  37   ch  China  ...        7.53            0.44       1367485388

[1 rows x 12 columns]

In [11]:
# Write a single query that returns minimum polution excluding the row for the whole world
min_population = pd.read_sql_query('''SELECT *, MIN(population) 
                                        FROM facts
                                       WHERE name <> 'World';''', conn)
min_population

id code        name  ... death_rate  migration_rate MIN(population)
0  250   ay  Antarctica  ...       None            None               0

[1 rows x 12 columns]

In [12]:
# Calculate the average value for the following columns:
avg_population = pd.read_sql_query(''' SELECT AVG(population)
                                         FROM facts
                                        WHERE name <> 'World';''', conn)
avg_population

AVG(population)
0     3.224267e+07

In [13]:
# Calculate the average value for the following columns:
avg_area = pd.read_sql_query(''' SELECT AVG(area)
                                         FROM facts
                                        WHERE name <> 'World';''', conn)
avg_area

AVG(area)
0  555093.546185

# Finding Densely Populated Countries


We'll identify countries that have the following:

* Above-average values for population.
* Below-average values for area.

In [24]:
# Write a single query return countries have population > avg_population
above_avg_population = pd.read_sql_query(''' SELECT *
                                                FROM facts
                                               WHERE name <> 'World' 
                                                 AND population > (SELECT AVG(population)
                                                                     FROM facts);''', conn)
above_avg_population

id code  ... death_rate  migration_rate
0    14   bg  ...       5.61            0.46
1    24   br  ...       6.58            0.14
2    37   ch  ...       7.53            0.44
3    40   cg  ...      10.07            0.27
4    53   eg  ...       4.77            0.19
5    58   et  ...       8.19            0.22
6    61   fr  ...       9.16            1.09
7    65   gm  ...      11.42            1.24
8    77   in  ...       7.32            0.04
9    78   id  ...       6.37            1.16
10   79   ir  ...       5.94            0.07
11   85   ja  ...       9.51            0.00
12  114   mx  ...       5.26            1.68
13  129   ni  ...      12.90            0.22
14  132   pk  ...       6.49            1.54
15  138   rp  ...       6.11            2.09
16  143   rs  ...      13.69            1.69
17  173   th  ...       7.80            0.00
18  179   tu  ...       5.88            2.16
19  185   uk  ...       9.35            2.54
20  186   us  ...       8.15            3.86
21  192   vm  ...       5.93            0.30
22  197   ee  ...      10.20            2.50

[23 rows x 11 columns]

In [23]:
# Write a single query return countries have population < avg_population
below_avg_population = pd.read_sql_query(''' SELECT *
                                                FROM facts
                                               WHERE name <> 'World' 
                                                 AND population < (SELECT AVG(population)
                                                                     FROM facts);''', conn)
below_avg_population

id code            name  ...  birth_rate  death_rate  migration_rate
0      1   af     Afghanistan  ...       38.57       13.89            1.51
1      2   al         Albania  ...       12.92        6.58            3.30
2      3   ag         Algeria  ...       23.67        4.31            0.92
3      4   an         Andorra  ...        8.13        6.96            0.00
4      5   ao          Angola  ...       38.78       11.49            0.46
..   ...  ...             ...  ...         ...         ...             ...
213  247   vq  Virgin Islands  ...       10.31        8.54            7.67
214  250   ay      Antarctica  ...         NaN         NaN             NaN
215  251   gz      Gaza Strip  ...       31.11        3.04            0.00
216  254   we       West Bank  ...       22.99        3.50            0.00
217  255   wi  Western Sahara  ...       30.24        8.34             NaN

[218 rows x 11 columns]

In [22]:
# Write a single query return countries have population > avg_area
above_avg_area = pd.read_sql_query(''' SELECT *
                                         FROM facts
                                        WHERE name <> 'World' 
                                          AND population > (SELECT AVG(area)
                                                                     FROM facts);''', conn)
above_avg_area

id code            name  ...  birth_rate  death_rate  migration_rate
0      1   af     Afghanistan  ...       38.57       13.89            1.51
1      2   al         Albania  ...       12.92        6.58            3.30
2      3   ag         Algeria  ...       23.67        4.31            0.92
3      5   ao          Angola  ...       38.78       11.49            0.46
4      7   ar       Argentina  ...       16.64        7.33            0.00
..   ...  ...             ...  ...         ...         ...             ...
168  205   mc           Macau  ...        8.88        4.22            3.37
169  246   rq     Puerto Rico  ...       10.86        8.67            8.15
170  251   gz      Gaza Strip  ...       31.11        3.04            0.00
171  254   we       West Bank  ...       22.99        3.50            0.00
172  255   wi  Western Sahara  ...       30.24        8.34             NaN

[173 rows x 11 columns]

In [21]:
# Write a single query return countries have population < avg_area
below_avg_area = pd.read_sql_query(''' SELECT *
                                         FROM facts
                                        WHERE name <> 'World' 
                                          AND population < (SELECT AVG(area)
                                                                     FROM facts);''', conn)
below_avg_area

id code                      name  ...  birth_rate  death_rate  migration_rate
0     4   an                   Andorra  ...        8.13        6.96            0.00
1     6   ac       Antigua and Barbuda  ...       15.85        5.69            2.21
2    12   bf              Bahamas, The  ...       15.50        7.05            0.00
3    15   bb                  Barbados  ...       11.87        8.44            0.30
4    18   bh                    Belize  ...       24.68        5.97            0.00
..  ...  ...                       ...  ...         ...         ...             ...
63  242   aq            American Samoa  ...       22.89        4.75           21.13
64  243   gq                      Guam  ...       16.82        5.12            6.34
65  245   cq  Northern Mariana Islands  ...       18.32        3.71            7.16
66  247   vq            Virgin Islands  ...       10.31        8.54            7.67
67  250   ay                Antarctica  ...         NaN         NaN             NaN

[68 rows x 11 columns]